### PR TITLE
Prototype of Hugging Face connector and extracting metadata from pipeline decorator

### PR DESCRIPTION
**Pull request by WillEngler**
Implements #291 

## Overview

Adds HFConnector class - the first "model connector". This provides a wrapper on Hugging Face Hub and lets you link your model to a pipeline.

Example usage:

```
from garden_ai.model_connectors import HFConnector
model_connector = HFConnector("willengler-uc/iris-classifier")

@garden_pipeline(model_connectors=[model_connector])
def classify_iris(iris_features, **kwargs):
    model_dir = model_connector.stage()
    iris_classifier = joblib.load(f"{model_dir}/iris_classifier_model.pkl")
    pred = iris_classifier.predict(iris_features)
    iris_classes = ['setosa', 'versicolor', 'virginica']
    return iris_classes[pred[0]]
```

This also adds a `garden-ai prototype extract` command as a proof of concept for how to pull metadata from a pipeline in a notebook.

## Discussion

The `extract` command probably won't last long - that should get folded into the publication step.

## Testing

I tested this locally with an example sklearn iris classifier notebook. I confirmed that I could pull down my HF model with the connector and use it in the pipeline. I further confirmed that I can get the ModelMetadata object from the container up to the host process.

## Documentation

🤠 

----
:books: Documentation preview :books:: https://garden-ai--298.org.readthedocs.build/en/298/

### CODE REVIEW

File: garden_ai/app/prototype.py [hunk 1]
```diff
@@ -158,6 +164,66 @@ def plant(
...
+    # Leading 4 spaces is because the first line wasn't being indented
+    function_body_dedented = textwrap.dedent("    " + function_body)
+
+    # Write the Python script to a temporary file on the host
+    with tempfile.NamedTemporaryFile(delete=False, suffix=".py") as f:
+        f.write(function_body_dedented.encode())
+        temp_file_name = f.name
+
+    mount_file = f"{temp_file_name}:/tmp/script.py"
+    interpreter_cmd = "python /tmp/script.py"
```
[the remark this answers is not in the document]
Comment (OwenPriceSkelly) [reply garden_ai/app/prototype.py:185]: could also skip the tempfile: 
```suggestion
    interpreter_cmd = f"python3 -c '{function_body_dedented}'"
```

File: garden_ai/container/containerize.py [hunk 2]
```diff
@@ -113,14 +119,24 @@ def start_container(
...
-    subprocess.run(run_command)
+    output = subprocess.run(run_command, capture_output=True, text=True)
+
     if cleanup:
         subprocess.run(["docker", "rmi", image_name])
 
+    return output.stdout
+
+
+def garden_pipeline(model_connectors=None):
```
Comment (OwenPriceSkelly) [inline garden_ai/container/containerize.py:130]: ooh this is one of my favorite python footguns: 
```
# both of these work the same, `myfunc` gets bound to the decorated function 
@garden_pipeline(model_connectors=None)
@garden_pipeline() 
def myfunc(): 
    ...
# breaks silently, bc `myfunc` gets bound to `garden_pipeline.<locals>.decorator`
@garden_pipeline 
def myfunc():
    ... 
```
check out the pattern I used for the `step` decorator, that's the least-kludgy way I know of getting `@foo()` and `@foo` to behave consistently

**Review by OwenPriceSkelly — APPROVED** [review]
🤠lgtm!🤠

some nits but nothing blocking

File: pyproject.toml [hunk 3]
```diff
@@ -37,6 +37,7 @@ typer = { extras = ["all"], version = "^0.7.0" }
 beartype = "^0.12.0"
 jinja2 = "^3.1.2"
 mlflow = "^2.5.0"
+huggingface-hub = "0.17.3"
```
[the remark this answers is not in the document]
Comment (WillEngler) [reply pyproject.toml:40]: Cool, will do

**Pull request merged**